packages/rpc-subscriptions/src/rpc-subscriptions-channel-pool.ts

```diff
@@ -0,0 +1,98 @@
+import { RpcSubscriptionsChannelCreator } from '@solana/rpc-subscriptions-spec';
+
+import { ChannelPoolEntry, createChannelPool } from './rpc-subscriptions-channel-pool-internal';
+
+type Config<TChannelCreator> = {
+    createChannel: TChannelCreator;
+    maxSubscriptionsPerChannel: number;
+    minChannels: number;
+};
+
+export function getChannelPoolingChannelCreator<
+    TChannelCreator extends RpcSubscriptionsChannelCreator<unknown, unknown>,
+>({ createChannel, maxSubscriptionsPerChannel, minChannels }: Config<TChannelCreator>): TChannelCreator {
+    const pool = createChannelPool();
+    /**
+     * This function advances the free channel index to the pool entry with the most capacity. It
+     * sets the index to `-1` if all channels are full.
+     */
+    function recomputeFreeChannelIndex() {
+        if (pool.entries.length < minChannels) {
+            // Don't set the free channel index until the pool fills up; we want to keep creating
+            // channels before we start rotating among them.
+            pool.freeChannelIndex = -1;
+            return;
+        }
+        let mostFreeChannel: Readonly<{ poolIndex: number; subscriptionCount: number }> | undefined;
+        for (let ii = 0; ii < pool.entries.length; ii++) {
+            const nextPoolIndex = (pool.freeChannelIndex + ii + 2) % pool.entries.length;
+            const nextPoolEntry =
+                // Start from the item two positions after the current item. This way, the
+                // search will finish on the item after the current one. This ensures that, if
+                // any channels tie for having the most capacity, the one that will be chosen is
+                // the one immediately to the current one's right (wrapping around).
```

Comment:
It's pretty insane, but in this case:

```
A(5), B(5), C(1), D(1) where A is the current one. Then here D is the next one whereas we were probably expecting C as per the rotation principle.
```

`A` can't be the current one, because the next-free-channel is chosen ahead of time. The next free channel would always be `C` or `D` depending on where we were in the rotation.

Next free channels are chosen:

* as a last step, after one is added.
* as a last step, after one is removed.

The next time that a channel is added, it goes right where the `freeChannelIndex` points, no questions asked.

I _think_ covered all the weird cases in the tests, but let me know if you can think of one that breaks things!

Reply:
If I understand correctly, the second scenario cannot happen due to how this function is being used within the decorator.

I think that's fine since we've got good test coverage but wouldn't it be better for maintainability if this function alone was responsible for keeping all the invariants of finding the next channel regardless of how it is being called?

Reply:
Well, I don't know! This is the way that made sense to me at the time. I feel like if you compute the free channel as-needed rather than ahead-of-time you would need to add some tracking variables to know what the ‘current’ channel is and you might end up with more code in the end, but ¯\\_(°ペ)\_/¯